internal/db/bootstrap_logs.go

```diff
@@ -0,0 +1,91 @@
+// Copyright 2025 Canonical.
+
+package db
+
+import (
+	"context"
+
+	"github.com/google/uuid"
+
+	"github.com/canonical/jimm/v3/internal/dbmodel"
+	"github.com/canonical/jimm/v3/internal/errors"
+	"github.com/canonical/jimm/v3/internal/servermon"
+)
+
+// AddBootstrapLog creates a bootstrap log entry.
+// It returns an error if the parameters cannot be validated or the database encounters an error.
+func (d *Database) AddBootstrapLog(ctx context.Context, jobId uuid.UUID, lineNumber int, logLine string) (err error) {
+	const op = errors.Op("db.CreateBootstrapLog")
+
+	log, err := dbmodel.NewBootstrapLog(jobId, lineNumber, logLine)
+	if err != nil {
+		return errors.E(op, "failed to construct bootstrap log", err)
+	}
+
+	if err := d.ready(); err != nil {
+		return errors.E(op, err)
+	}
+
+	durationObserver := servermon.DurationObserver(servermon.DBQueryDurationHistogram, string(op))
+	defer durationObserver()
+	defer servermon.ErrorCounter(servermon.DBQueryErrorCount, &err, string(op))
+
+	if err := d.DB.WithContext(ctx).Create(log).Error; err != nil {
+		return errors.E(op, dbError(err))
+	}
+
+	return nil
+}
+
+// QueryBootstrapLog queries for bootstrap logs based on the jobId and offset.
+// It returns an error if the job doesn't exist, there aren't any logs or the offset
+// is greater than the amount of logs,
+func (d *Database) QueryBootstrapLog(ctx context.Context, jobId uuid.UUID, offset int) (loggies []string, err error) {
+	const op = errors.Op("db.QueryBootstrapLog")
+
+	if err := d.ready(); err != nil {
+		return loggies, errors.E(op, err)
+	}
+
+	durationObserver := servermon.DurationObserver(servermon.DBQueryDurationHistogram, string(op))
+	defer durationObserver()
+	defer servermon.ErrorCounter(servermon.DBQueryErrorCount, &err, string(op))
+
+	// Make sure job exists, if it doesn't, there's no point running the query
+	if err := d.DB.WithContext(ctx).First(&dbmodel.JobTrackerEntry{JobID: jobId}, "job_id = ?", jobId).Error; err != nil {
+		return loggies, errors.E(op, "job not found", dbError(err))
+	}
```

Comment:
We can let the DB handle this, no need to do a check that the SQL will handle right?

Reply:
Well db will just say log not found, not "job not found" when we do a select. I wanted a nice explicit error?